qiime2/sdk/asynchronous_context.py

```diff
@@ -0,0 +1,55 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2023, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import concurrent.futures
+
+from qiime2.sdk.context import Context
+from qiime2.sdk.serial_context import SerialContext
+
+
+def _subprocess_apply(ctx, args, kwargs):
+    exe = ctx.action_obj._bind(
+        lambda: SerialContext(ctx), {'type': 'asynchronous'})
+    results = exe(*args, **kwargs)
+
+    return results
+
+
+class AsynchronousContext(Context):
+    def dispatch(self, *args, **kwargs):
+        # TODO handle this better in the future, but stop the massive error
+        # caused by MacOSX asynchronous runs for now.
+        try:
+            import matplotlib as plt
+            if plt.rcParams['backend'].lower() == 'macosx':
+                raise EnvironmentError(backend_error_template %
+                                       plt.matplotlib_fname())
+        except ImportError:
+            pass
+
+        # This function's signature is rewritten below using
+        # `decorator.decorator`. When the signature is rewritten, args[0]
+        # is the function whose signature was used to rewrite this
+        # function's signature.
+        args = args[1:]
+
+        pool = concurrent.futures.ProcessPoolExecutor(max_workers=1)
+        future = pool.submit(_subprocess_apply, self, args, kwargs)
+        # TODO: pool.shutdown(wait=False) caused the child process to
+        # hang unrecoverably. This seems to be a bug in Python 3.7
+        # It's probably best to gut concurrent.futures entirely, so we're
+        # ignoring the resource leakage for the moment.
+        return future
```

Comment:
Does this work? I would assume we would need a proxy here. Or otherwise, have some mechanism to roll the entire pipeline/callable into the single dispatch.

Reply:
Idk we never really supported async much, so I wasn't sure how far we wanted to go with it.

My understanding based on reading the implementation and talking to you about it was that it didn't actually parallelize anything and we didn't care for it to parallelize anything.

It was more of an "I'll execute it when I execute it" type thing. Idk.

Reply:
Do we not test async pipelines? Blocking within (if not root) is fine, but I kind of expect a pipeline to break since `_dispatch_` lives inside the center of the returned callable, and it wouldn't expect a future in those circumstances.

Reply:
Nevermind. Child contexts are synchronous, so that's why this works fine. In which case, this is perfectly fine and matches the old API.